### DISCUSSION TITLE
OpenTelemetry tracing support?

### DISCUSSION THREAD

**Discussion by stefee** [nodejs/undici#1439] [Q&A]
Does undici work out of the box with node auto-instrumentation for OTEL? https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/metapackages/auto-instrumentations-node

The auto instrumentation includes shims for the node core modules `node:http`, `node:net`, etc.

It seems that trace propagation across service boundaries will only work for libraries that are using `http.request` under the hood (e.g. node-fetch). Does undici use `http.request` under the hood or does it use something more primitive like `tcp.connect` instead? This might prevent the OTEL instrumentation from injecting the headers required for traces to get propagated?

I also discovered that this package exists, but it might make sense to provide an "official" integration or recommendation: https://github.com/gadget-inc/opentelemetry-instrumentations

Update: Twitter user confirmed to me that it doesn’t work out of the box with the auto instrumentation
https://twitter.com/isntitvacant/status/1524869221280190464?s=21&t=PquRXpxyYTBYaDlk2uA9cA

**Comment by mcollina** [comment]
Correct. Would you like to help and implement this? It would need to use https://github.com/nodejs/undici/blob/main/docs/api/DiagnosticsChannel.md to implement everything and restrain from massive monkeypatching. We will be happy to respond to any questions.

**Reply by stefee** [reply]
Yeah, happy to help. Need to check which Node.js versions the OTEL library expects to support - v16+ might not be good enough.

**Reply by metcoder95** [reply]
Out of curiosity, DiagnosticChannel will be sufficient for the purpose of instrumenting Undici to OTEL compliance? I think is mostly made for reporting right? But the issue will come at the propagation level, as in order to propagate the trace ids through the headers the request must be altered on-the-flight, which is what the unofficial lib is doing (see [here](https://github.com/gadget-inc/opentelemetry-instrumentations/blob/79ed697b56aa4a008374e317986164ee3fa97326/packages/opentelemetry-instrumentation-undici/src/index.ts#L62-L125)).

Not so sure if the API lifecycle events are enough, but maybe worth it to try. I think this PR opens some similar surface #1338 

Happy to also jump in and help 🙂

**Reply by ronag** [reply]
Looks to me like OTEL already has undici support? What's missing?

**Reply by stefee** [reply]
I'm also not sure if DiagnosticChannel will be enough - it depends if it allows to inject the propagation headers from the trace context on the request?

@ronag Can you link to it here, please? OTEL supports the core http/https modules, but since undici does its own implementation of http on top of net/tcp modules, OTEL has no way of injecting propagation headers on http reqs.

**Reply by ronag** [reply]
https://github.com/gadget-inc/opentelemetry-instrumentations/blob/79ed697b56aa4a008374e317986164ee3fa97326/packages/opentelemetry-instrumentation-undici/src/index.ts#L62-L125

**Reply by stefee** [reply]
OK I see what you’re saying now - yes, changes to undici might not be required in order for OTEL to support it if we’re happy with the monkey-patch approach.

**Reply by stefee** [reply]
If we’re happy to use this method I’d be happy to try and land this in the contrib repo as part of the “auto” instrumentation: https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/metapackages/auto-instrumentations-node see relevant issue https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1021

**Reply by mcollina** [reply]
I would prefer if we used diagnostics-channel as it's significantly faster than monkey-patching.

The ALS context is already propagated automatically with Undici, you don't have to worry about that.

**Reply by stefee** [reply]
To confirm @mcollina is the idea then to use diagnostics-channel to create an OTEL instrumentation in the opentelemetry-js-contrib repo? i.e. similar to the gadget-inc instrumentation, but just using diagnostics-channel rather than patching.

**Reply by dyladan** [reply]
Hello, OTel JS maintainer here. I'm quite excited about the idea of a diagnostics channel based instrumentation as I haven't seen one yet. I am only concerned we might miss the propagation. If uncici emits an event that a request is started and OTel starts a new span, how is the span context added to the traceparent header? I'm definitely willing to help facilitate this on the OTel end.

**Reply by mcollina** [reply]
We use AsyncResource internally. If something is not propagated, we'll fix it here.

**Reply by stefee** [reply]
I’m also struggling to understand how undici is able to get the span context from OTEL and add the propagation headers on the request.

**Reply by stefee** [reply]
For example, injecting `X-B3-*` headers - this needs to happen somewhere. https://github.com/openzipkin/b3-propagation

It doesn’t make sense for undici to be aware of how this works. It doesn’t look like diagnostics provides a way to inject headers.

**Reply by stefee** [reply]
Hang on a sec - I take that back. I should learn to read. https://github.com/nodejs/undici/blob/main/docs/api/DiagnosticsChannel.md#undicirequestcreate

```js
request.addHeader('hello', 'world')
```

Should give us what we need.

**Reply by stefee** [reply]
I think we were crossing wires @mcollina - OTEL context propagation is _not_ handled automatically by undici, it needs to be handled by the instrumentation by injecting headers. But this _can_ be achieved using diag channels rather than needing to monkey-patch, you’re right.

**Reply by mcollina** [reply]
monkey-patching is (mostly) needed because libraries can't maintain the asynchronous context across callbacks and promise chains. The problem is not to add headers in the right moment, but rather make sure the information to put there is available. This is a problem for a lot of libraries. We commit to fully support async_hooks and ALS in undici.

**Reply by stefee** [reply]
I’m looking at OTEL in more detail now and will update here https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1021

**Reply by Flarna** [reply]
I think from OTel point of view the use of [TracingChannel](https://nodejs.org/dist/latest-v21.x/docs/api/diagnostics_channel.html#class-tracingchannel) would be a better fit instead of simple channels. It avoid the need of an [enterWith](https://nodejs.org/dist/latest-v21.x/docs/api/async_context.html#asynclocalstorageenterwithstore) like API.

An `enterWith()` like API is not part of the TC39 proposal for [`AsyncContext`](https://github.com/tc39/proposal-async-context) and therefore likely not future proof.
